src/Microsoft.DotNet.TestFramework/TestAssetInfo.cs

```diff
@@ -242,5 +298,66 @@ private void DoBuild()
                 throw new Exception(message);
             }
         }
+
+        private void ThrowIfAssetSourcesHaveChanged()
+        {
+            if (!_dataDirectory.Exists)
+            {
+                return;
+            }
+
+            var dataDirectoryFiles = _dataDirectory.GetFiles("*", SearchOption.AllDirectories);
+
+            if (!dataDirectoryFiles.Any())
+            {
+                return;
+            }
+
+            var trackedFiles = _inventoryFiles.AllInventoryFiles.SelectMany(f => LoadInventory(f));
+
+            var assetFiles = GetFileList();
+
+            var untrackedFiles = assetFiles.Where(a => !trackedFiles.Any(t => t.FullName.Equals(a.FullName)));
+
+            if (untrackedFiles.Any())
+            {
+                var message = $"TestAsset {_assetName} has untracked files." +
+                    "Consider cleaning the asset and deleting its `.tam` directory to " + 
+                    "recreate tracking files.\n\n" +
+                    $".tam directory: {_dataDirectory.FullName}\n" +
+                    "Untracked Files: \n";
+
+                message += String.Join("\n", untrackedFiles.Select(f => $" - {f.FullName}\n"));
+
+                throw new Exception(message);
+            }
+
+            var earliestDataDirectoryTimestamp =
+                dataDirectoryFiles
+                    .OrderBy(f => f.LastWriteTime)
+                    .First()
+                    .LastWriteTime;
+
+            if (earliestDataDirectoryTimestamp == null)
+            {
+                return;
+            }
+
+            var updatedSourceFiles = LoadInventory(_inventoryFiles.Source)
+                .Where(f => f.LastWriteTime > earliestDataDirectoryTimestamp);
+
+            if (updatedSourceFiles.Any())
+            {
+                var message = $"TestAsset {_assetName} has updated files." +
+                    "Consider cleaning the asset and deleting its `.tam` directory to " + 
+                    "recreate tracking files.\n\n" +
+                    $".tam directory: {_dataDirectory.FullName}\n" +
+                    "Updated Files: \n";
+
+                message += String.Join("\n", updatedSourceFiles.Select(f => $" - {f.FullName}\n"));
+
+                throw new Exception(message);
```

Comment:
This is a pattern I'd love to see rolled back. A good rule of thumb is to never throw `Exception` (because you can only catch it by catching `Exception` but you should only ever catch exceptions you can actually handle).